### PR TITLE
Upgrade Contrib  to v0.103.0

### PR DESCRIPTION
**Pull request by okankoAMZ**
# Description of the issue
Contrib was last updated 2 months ago. Contrib repository PR can be found [here](https://github.com/amazon-contributing/opentelemetry-collector-contrib/pull/223). Please do not merge before that PR.

# Description of changes
Upgrade contrib to v0.103.0 stable version. To do that I had do: 
* Update CloudWatch agent command, where I am using now passing in the Provider Settings and not just Provider. 
* Update Config Provider because the inputs have changed to Factories and Settings from directly feeding in the Converters and Providers.
* Update our logger, so when a NewLogger is created now there is a option to retrieve the Logger Options with the Logger itself. 
* `component.UnmarshalConfig` was deprecated so I had to migrate it to `confmap.Unmarshal`
* Updated translator sample configs.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Ran integration test here: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/10063006520
More recent integration test here:https://github.com/aws/amazon-cloudwatch-agent/actions/runs/10101543388

Latest agent logs:
```
[ec2-user@ip-172-31-46-210 amazon-cloudwatch-agent]$ tail -n 100 /opt/aws/amazon-cloudwatch-agent/logs/amazon-cloudwatch-agent.log 
        logs:
            development: false
            disable_caller: false
            disable_stacktrace: false
            encoding: console
            level: debug
            output_paths:
                - /opt/aws/amazon-cloudwatch-agent/logs/amazon-cloudwatch-agent.log
            sampling:
                enabled: true
                initial: 2
                thereafter: 500
                tick: 10s
        metrics:
            address: ""
            level: None
        traces: {}
2024/08/05 18:27:58 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json ...
2024/08/05 18:27:58 I! Valid Json input schema.
2024/08/05 18:27:58 I! Detected runAsUser: root
2024/08/05 18:27:58 I! Changing ownership of [/opt/aws/amazon-cloudwatch-agent/logs /opt/aws/amazon-cloudwatch-agent/etc /opt/aws/amazon-cloudwatch-agent/var] to 0:0
2024-08-05T18:27:58Z I! Starting AmazonCloudWatchAgent CWAgent/1.300041.1 (go1.22.4; linux; amd64) with log file /opt/aws/amazon-cloudwatch-agent/logs/amazon-cloudwatch-agent.log with log target lumberjack
2024-08-05T18:27:58Z I! AWS SDK log level not set
2024-08-05T18:27:58Z I! creating new logs agent
2024-08-05T18:27:58Z I! [logagent] starting
2024-08-05T18:27:58Z I! {"caller":"service@v0.98.0/telemetry.go:47","msg":"Skipping telemetry setup.","address":"","level":"None"}
2024-08-05T18:27:58Z D! {"caller":"extension@v0.98.0/extension.go:165","msg":"Alpha component. May change in the future.","kind":"extension","name":"agenthealth/metrics"}
2024-08-05T18:27:58Z D! {"caller":"exporter@v0.98.0/exporter.go:273","msg":"Alpha component. May change in the future.","kind":"exporter","data_type":"metrics","name":"awscloudwatch"}
2024-08-05T18:27:58Z D! {"caller":"receiver@v0.98.0/receiver.go:308","msg":"Stable component.","kind":"receiver","name":"telegraf_mem","data_type":"metrics"}
2024-08-05T18:27:58Z I! {"caller":"service@v0.98.0/service.go:143","msg":"Starting CWAgent...","Version":"1.300041.1","NumCPU":2}
2024-08-05T18:27:58Z I! {"caller":"extensions/extensions.go:34","msg":"Starting extensions..."}
2024-08-05T18:27:58Z I! {"caller":"extensions/extensions.go:37","msg":"Extension is starting...","kind":"extension","name":"agenthealth/metrics"}
2024-08-05T18:27:58Z I! {"caller":"extensions/extensions.go:52","msg":"Extension started.","kind":"extension","name":"agenthealth/metrics"}
2024-08-05T18:27:58Z D! Fallback shared config file(s): [.aws/credentials]
2024-08-05T18:27:58Z D! Successfully created credential sessions
2024-08-05T18:27:58Z E! Failed to get credential from session: NoCredentialProviders: no valid providers in chain
caused by: EnvAccessKeyNotFound: failed to find credentials in the environment.
SharedCredsLoad: failed to load profile, .
EC2RoleRequestError: no EC2 instance role found
caused by: EC2MetadataError: failed to make EC2Metadata request
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
                 "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>404 - Not Found</title>
 </head>
 <body>
  <h1>404 - Not Found</h1>
 </body>
</html>

        status code: 404, request id: 
2024-08-05T18:27:58Z D! {"caller":"awsmiddleware@v0.0.0-20240604205355-82dc97164e88/helper.go:18","msg":"Configured middleware on AWS client","kind":"exporter","data_type":"metrics","name":"awscloudwatch","middleware":"agenthealth/metrics"}
2024-08-05T18:27:58Z I! cloudwatch: get unique roll up list []
2024-08-05T18:27:58Z D! {"caller":"adapter/receiver.go:41","msg":"Starting adapter","kind":"receiver","name":"telegraf_mem","data_type":"metrics","receiver":"mem"}
2024-08-05T18:27:58Z I! {"caller":"service@v0.98.0/service.go:169","msg":"Everything is ready. Begin running and processing data."}
2024-08-05T18:27:58Z W! {"caller":"localhostgate/featuregate.go:63","msg":"The default endpoints for all servers in components will change to use localhost instead of 0.0.0.0 in a future version. Use the feature gate to preview the new default.","feature gate ID":"component.UseLocalHostAsDefaultHost"}
2024-08-05T18:27:58Z I! cloudwatch: publish with ForceFlushInterval: 1m0s, Publish Jitter: 29.991730558s
2024-08-05T18:27:59Z D! [logagent] open file count, 0
2024-08-05T18:27:59Z D! {"caller":"adapter/receiver.go:61","msg":"Begin scraping metrics with adapter","kind":"receiver","name":"telegraf_mem","data_type":"metrics","receiver":"mem"}
2024-08-05T18:28:00Z D! [logagent] open file count, 0
2024-08-05T18:28:01Z D! [logagent] open file count, 0
2024-08-05T18:28:02Z D! [logagent] open file count, 0
2024-08-05T18:28:03Z D! [logagent] open file count, 0
2024-08-05T18:28:04Z D! [logagent] open file count, 0
2024-08-05T18:28:05Z D! [logagent] open file count, 0
2024-08-05T18:28:06Z D! [logagent] open file count, 0
2024-08-05T18:28:07Z D! [logagent] open file count, 0
2024-08-05T18:28:08Z D! [logagent] open file count, 0
2024-08-05T18:28:09Z D! [logagent] open file count, 0
2024-08-05T18:28:10Z D! [logagent] open file count, 0
2024-08-05T18:28:11Z D! [logagent] open file count, 0
2024-08-05T18:28:12Z D! [logagent] open file count, 0
2024-08-05T18:28:13Z D! [logagent] open file count, 0
2024-08-05T18:28:14Z D! [logagent] open file count, 0
2024-08-05T18:28:15Z D! [logagent] open file count, 0
2024-08-05T18:28:16Z D! [logagent] open file count, 0
2024-08-05T18:28:17Z D! [logagent] open file count, 0
2024-08-05T18:28:18Z D! [logagent] open file count, 0
2024-08-05T18:28:19Z D! [logagent] open file count, 0
2024-08-05T18:28:20Z D! [logagent] open file count, 0
2024-08-05T18:28:21Z D! [logagent] open file count, 0
2024-08-05T18:28:22Z D! [logagent] open file count, 0
2024-08-05T18:28:23Z D! [logagent] open file count, 0
2024-08-05T18:28:24Z D! [logagent] open file count, 0
2024-08-05T18:28:25Z D! [logagent] open file count, 0
2024-08-05T18:28:26Z D! [logagent] open file count, 0
2024-08-05T18:28:27Z D! [logagent] open file count, 0
2024-08-05T18:28:28Z D! [logagent] open file count, 0
2024-08-05T18:28:28Z I! CWAGENT_LOG_LEVEL is set to "DEBUG"
``` 
after my changes:
```
 2024-08-05T20:02:44Z D! Using credential [REDACTED] from EC2RoleProvider
2024-08-05T20:02:44Z I! {"caller":"service@v0.103.0/service.go:182","msg":"Starting CWAgent...","Version":"1.300042.1-32-ge7225958-untracked","NumCPU":2}
2024-08-05T20:02:44Z I! {"caller":"extensions/extensions.go:34","msg":"Starting extensions..."}
2024-08-05T20:02:44Z I! {"caller":"extensions/extensions.go:37","msg":"Extension is starting...","kind":"extension","name":"agenthealth/metrics"}
2024-08-05T20:02:44Z I! {"caller":"extensions/extensions.go:52","msg":"Extension started.","kind":"extension","name":"agenthealth/metrics"}
2024-08-05T20:02:44Z D! Fallback shared config file(s): [/home/cwagent/.aws/credentials]
2024-08-05T20:02:44Z D! Successfully created credential sessions
2024-08-05T20:02:44Z D! Using credential [REDACTED]from EC2RoleProvider
2024-08-05T20:02:44Z I! cloudwatch: get unique roll up list []
2024-08-05T20:02:44Z I! {"caller":"ec2tagger/ec2tagger.go:415","msg":"ec2tagger: Check EC2 Metadata.","kind":"processor","name":"ec2tagger","pipeline":"metrics/host"}
2024-08-05T20:02:44Z I! cloudwatch: publish with ForceFlushInterval: 1m0s, Publish Jitter: 5.078351757s
2024-08-05T20:02:44Z D! Fallback shared config file(s): [/home/cwagent/.aws/credentials]
2024-08-05T20:02:44Z D! Successfully created credential sessions
2024-08-05T20:02:44Z D! Using credential [REDACTED] from EC2RoleProvider
2024-08-05T20:02:44Z I! {"caller":"ec2tagger/ec2tagger.go:333","msg":"ec2tagger: EC2 tagger has started initialization.","kind":"processor","name":"ec2tagger","pipeline":"metrics/host"}
2024-08-05T20:02:44Z I! {"caller":"service@v0.103.0/service.go:208","msg":"Everything is ready. Begin running and processing data."}
2024-08-05T20:02:44Z W! {"caller":"localhostgate/featuregate.go:63","msg":"The default endpoints for all servers in components will change to use localhost instead of 0.0.0.0 in a future version. Use the feature gate to preview the new default.","feature gate ID":"component.UseLocalHostAsDefaultHost"}
2024-08-05T20:02:44Z W! {"caller":"ec2tagger/ec2tagger.go:465","msg":"ec2tagger: Unable to describe ec2 tags for initial retrieval","kind":"processor","name":"ec2tagger","pipeline":"metrics/host","error":"UnauthorizedOperation: You are not authorized to perform this operation. User: arn:aws:sts::956457624121:assumed-role/SSMDefaultRoleForPVREReporting/i-06aad6834bc794769 is not authorized to perform: ec2:DescribeTags because no identity-based policy allows the ec2:DescribeTags action\n\tstatus code: 403, request id: 0b615dc1-b942-4c17-b77d-84ffff632f4d"}
2024-08-05T20:02:45Z D! [logagent] open file count, 0
2024-08-05T20:02:46Z D! [logagent] open file count, 0
2024-08-05T20:02:47Z D! [logagent] open file count, 0
2024-08-05T20:02:48Z D! [logagent] open file count, 0
2024-08-05T20:02:49Z D! [logagent] open file count, 0
2024-08-05T20:02:50Z D! [logagent] open file count, 0
2024-08-05T20:02:51Z D! [logagent] open file count, 0
2024-08-05T20:02:52Z D! [logagent] open file count, 0
2024-08-05T20:02:53Z D! [logagent] open file count, 0
2024-08-05T20:02:54Z D! [logagent] open file count, 0
2024-08-05T20:02:55Z D! [logagent] open file count, 0
2024-08-05T20:02:56Z D! [logagent] open file count, 0
2024-08-05T20:02:57Z D! [logagent] open file count, 0
2024-08-05T20:02:58Z D! [logagent] open file count, 0
2024-08-05T20:02:59Z D! [logagent] open file count, 0
2024-08-05T20:03:00Z D! [logagent] open file count, 0
2024-08-05T20:03:01Z D! [logagent] open file count, 0
2024-08-05T20:03:02Z D! [logagent] open file count, 0
2024-08-05T20:03:03Z D! [logagent] open file count, 0
2024-08-05T20:03:04Z D! [logagent] open file count, 0
2024-08-05T20:03:05Z D! [logagent] open file count, 0
2024-08-05T20:03:06Z D! [logagent] open file count, 0
2024-08-05T20:03:07Z D! [logagent] open file count, 0
2024-08-05T20:03:08Z D! [logagent] open file count, 0
2024-08-05T20:03:09Z D! [logagent] open file count, 0
2024-08-05T20:03:10Z D! [logagent] open file count, 0
2024-08-05T20:03:11Z D! [logagent] open file count, 0
2024-08-05T20:03:12Z D! [logagent] open file count, 0
2024-08-05T20:03:13Z D! [logagent] open file count, 0
2024-08-05T20:03:14Z I! CWAGENT_LOG_LEVEL is set to "DEBUG"
```

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
2. Run `make lint`

### CODE REVIEW

**Comment by ZhenyuTan-amz** [comment]
branch named with v102 but eventually bumped to v103. Is it just us trying to pulling in whatever is the latest? (just curious if it's fully intended, not against it at all)

**Comment by okankoAMZ** [comment]
> branch named with v102 but eventually bumped to v103. Is it just us trying to pulling in whatever is the latest? (just curious if it's fully intended, not against it at all)

Initially it we were migrating to v102 but due to parity issue between v102.1 and v102.0 we have decided to bump to v103. I renamed the PR but did not recreate the branch

File: cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go [hunk 1]
```diff
@@ -323,9 +319,11 @@ func runAgent(ctx context.Context,
 		}
 	}
 	// Else start OTEL and rely on adapter package to start the logfile plugin.
-
 	yamlConfigPath := *fOtelConfig
-	provider, err := configprovider.Get(yamlConfigPath)
+	level := cwaLogger.ConvertToAtomicLevel(wlog.LogLevel())
+	logger, loggerOptions := cwaLogger.NewLogger(writer, level)
+	providerSettings, err := configprovider.GetSettings(yamlConfigPath, logger)
+	provider, err := configprovider.Get(yamlConfigPath, logger)
```
Comment (jefchien) [inline cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go:326]: Create the provider from the settings.

**Review by musa-asad — DISMISSED** [review]

File: cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go [hunk 2]
```diff
@@ -341,31 +339,28 @@ func runAgent(ctx context.Context,
...
-	// unrecognized args. So below change overwrites the args. Need to investigate this further as I dont think the config
-	// path below here is actually used and it still respects what was set in the settings above.
-	e := []string{"--config=" + yamlConfigPath + " --feature-gates=exporter.xray.allowDot"}
+	// *************************************************************************************************
+	// ⚠️ WARNING ⚠️
+	// Noticed that args of parent process get passed here to otel collector which causes failures
+	// complaining about unrecognized args. So below change overwrites the args. Need to investigate this
+	// further as I don't think the config path below here is actually used and it still respects what
+	// was set in the settings above.
+	// *************************************************************************************************
```
[the remark this answers is not in the document]
Comment (jefchien) [reply cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go:352]: https://github.com/open-telemetry/opentelemetry-collector/blob/main/otelcol/command.go#L63

File: service/configprovider/provider.go [hunk 3]
```diff
@@ -8,18 +8,19 @@ import (
...
-			},
+			URIs:               []string{configPath},
+			ProviderFactories:  []confmap.ProviderFactory{fprovider},
+			ProviderSettings:   confmap.ProviderSettings{Logger: logger},
+			ConverterFactories: []confmap.ConverterFactory{expandconverter.NewFactory()},
+			ConverterSettings:  confmap.ConverterSettings{Logger: logger},
 		},
 	}
-	return otelcol.NewConfigProvider(settings)
+	return settings, nil
```
Comment (jefchien) [inline service/configprovider/provider.go:25]: Why does `GetSettings` return an error if this is always nil?

File: translator/tocwconfig/sampleConfig/advanced_config_darwin.yaml [hunk 4]
```diff
@@ -72,11 +72,11 @@ service:
             processors:
                 - ec2tagger
             receivers:
-                - telegraf_mem
                 - telegraf_netstat
                 - telegraf_swap
                 - telegraf_cpu
                 - telegraf_disk
+                - telegraf_mem
```
Comment (jefchien) [inline translator/tocwconfig/sampleConfig/advanced_config_darwin.yaml:79]: Most of these didn't need to get re-generated. Makes it a bit harder to tell which configurations actually changed, but it's fine.

File: cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go [hunk 5]
```diff
@@ -341,31 +339,28 @@ func runAgent(ctx context.Context,
...
-
-	// Noticed that args of parent process get passed here to otel collector which causes failures complaining about
-	// unrecognized args. So below change overwrites the args. Need to investigate this further as I dont think the config
-	// path below here is actually used and it still respects what was set in the settings above.
-	e := []string{"--config=" + yamlConfigPath + " --feature-gates=exporter.xray.allowDot"}
+	// *************************************************************************************************
+	// ⚠️ WARNING ⚠️
+	// Noticed that args of parent process get passed here to otel collector which causes failures
+	// complaining about unrecognized args. So below change overwrites the args.
+	// The config path below here is actually used that was set in the settings above.
```
Comment (jefchien) [inline cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go:350]: nit: The wording here is a bit confusing.

**Review by sky333999 — APPROVED** [review]

File: cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go [hunk 6]
```diff
@@ -341,31 +339,28 @@ func runAgent(ctx context.Context,
...
+	// docs: https://github.com/open-telemetry/opentelemetry-collector/blob/main/otelcol/command.go#L63
```
Comment (jefchien) [inline cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go:351]: nit: Change to permalink. This line or file may be moved or renamed in other versions.

File: go.mod [hunk 7]
```diff
@@ -95,7 +90,7 @@ require (
 	github.com/BurntSushi/toml v1.3.2
 	github.com/Jeffail/gabs v1.4.0
 	github.com/amazon-contributing/opentelemetry-collector-contrib/extension/awsmiddleware v0.0.0-20240709194807-b0f0c0eda01b
```
Comment (jefchien) [inline go.mod:92]: Change this one and `github.com/amazon-contributing/opentelemetry-collector-contrib/override/aws` in the indirect require block as well to match the other contrib versions.

File: go.mod [hunk 8]
```diff
@@ -113,59 +108,59 @@ require (
...
+	go.opentelemetry.io/collector/config/configtelemetry v0.103.0
+	go.opentelemetry.io/collector/config/configtls v0.103.0
+	go.opentelemetry.io/collector/confmap v0.103.0
+	go.opentelemetry.io/collector/confmap/converter/expandconverter v0.103.0
+	go.opentelemetry.io/collector/confmap/provider/fileprovider v0.103.0
+	go.opentelemetry.io/collector/consumer v0.103.0
+	go.opentelemetry.io/collector/exporter v0.103.0
+	go.opentelemetry.io/collector/exporter/debugexporter v0.103.0
+	go.opentelemetry.io/collector/extension v0.103.0
+	go.opentelemetry.io/collector/featuregate v1.10.0 // indirect
```
Comment (jefchien) [inline go.mod:147]: nit: Move to indirect require block of go.mod below.

File: logger/logger.go [hunk 9]
```diff
@@ -20,6 +20,20 @@ type TelegrafWrapperEncoder struct {
...
+	)
+	option := zap.WrapCore(func(zapcore.Core) zapcore.Core {
+		return core
+	})
+	logger := zap.New(core)
+	defer logger.Sync()
+
+	return logger, []zap.Option{option}
+}
 func NewLoggerOptions(writer io.Writer, level zap.AtomicLevel) []zap.Option {
```
Comment (jefchien) [inline logger/logger.go:37]: Is this still used?

File: service/configprovider/provider.go [hunk 10]
```diff
@@ -8,18 +8,19 @@ import (
...
+	fprovider := fileprovider.NewFactory()
 	settings := otelcol.ConfigProviderSettings{
 		ResolverSettings: confmap.ResolverSettings{
-			URIs:       []string{configPath},
-			Converters: []confmap.Converter{expandconverter.New(confmap.ConverterSettings{})},
-			Providers: map[string]confmap.Provider{
-				fprovider.Scheme(): fprovider,
-			},
+			URIs:               []string{configPath},
+			ProviderFactories:  []confmap.ProviderFactory{fprovider},
```
Comment (jefchien) [inline service/configprovider/provider.go:19]: nit: No longer need the variable.
```suggestion
			ProviderFactories:  []confmap.ProviderFactory{fileprovider.NewFactory()},
```

**Review by jefchien — APPROVED** [review]

**Review by sky333999 — APPROVED** [review]

**Pull request merged**